util/types/field_type.go

```diff
@@ -91,7 +91,8 @@ func AggTypeClass(tps []*FieldType, flag *uint) TypeClass {
 			continue
 		}
 		argTypeClass := argFieldType.ToClass()
-		if argTypeClass == ClassString && mysql.HasBinaryFlag(argFieldType.Flag) {
+		if (IsTypeBlob(argFieldType.Tp) || IsTypeVarchar(argFieldType.Tp) || IsTypeChar(argFieldType.Tp)) &&
```

Comment:
borrowed from https://github.com/pingcap/tidb/pull/4158/commits/aa13a8fe674932ad0b9348e85010bc4e74a3f968

in pr https://github.com/pingcap/tidb/pull/4158

After it's merged I'll fix the conflicts